substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,756 @@
+	/// data and perform other sanity checks as the approvals map is calculated.
+	///
+	/// NOTE: this is an expensive state check since it iterates over all the nodes in the
+	/// target and voter list providers.
+	///
+	/// Invariants:
+	///
+	/// * Target List:
+	///   * The sum of the calculated approvals stake is the same as the current approvals in
+	///   the target list per target.
+	///   * The target score of an active validator is the sum of all of its nominators' stake
+	///   and the self-stake;
+	///   * The target score of an idle validator (i.e. chilled) is the sum of its nominator's
+	///   stake. An idle target may not be part of the target list, if it has no nominations.
+	///   * The target score of a "dangling" target (ie. idle AND unbonded validator) must
+	///   always be > 0. We expect the stake-tracker to have cleaned up dangling targets with 0
+	///   score.
+	///   * The number of target nodes in the target list matches the number of
+	///   (active_validators + idle_validators + dangling_targets_score_with_score).
+	///
+	/// * Voter List:
+	///  * The voter score is the same as the active stake of the corresponding stash.
+	///  * An active validator should also be part of the voter list.
+	///  * An idle validator should not be part of the voter list.
+	///  * A dangling target shoud not be part of the voter list.
+	pub(crate) fn do_try_state_approvals() -> Result<(), sp_runtime::TryRuntimeError> {
+		let mut approvals_map: BTreeMap<AccountIdOf<T>, sp_npos_elections::ExtendedBalance> =
+			BTreeMap::new();
+
+		// build map of approvals stakes from the `VoterList` POV.
+		for voter in T::VoterList::iter() {
+			if let Some(nominations) = <T::Staking as StakingInterface>::nominations(&voter) {
+				let score = <T::VoterList as SortedListProvider<AccountIdOf<T>>>::get_score(&voter)
+					.map_err(|_| "nominator score must exist in voter bags list")?;
+
+				// sanity check.
+				let active_stake = T::Staking::stake(&voter)
+					.map(|s| Self::weight_of(s.active))
+					.expect("active voter has bonded stake; qed.");
+				frame_support::ensure!(
+					active_stake == score,
+					"voter score must be the same as its active stake"
+				);
+
+				for nomination in nominations {
+					if let Some(stake) = approvals_map.get_mut(&nomination) {
+						*stake += score as sp_npos_elections::ExtendedBalance;
+					} else {
+						approvals_map.insert(nomination, score.into());
+					}
+				}
+			} else {
+				// if it is in the voter list but it's not a nominator, it should be a validator
+				// and part of the target list.
+				frame_support::ensure!(
+					T::Staking::status(&voter) == Ok(StakerStatus::Validator),
+					"wrong state of voter"
+				);
+				frame_support::ensure!(
+					T::TargetList::contains(&voter),
+					"if voter is in voter list and it's not a nominator, it must be a target"
+				);
+			}
+		}
+
+		// add self-vote of active targets to calculated approvals from the `TargetList` POV.
```

Comment:
Targets are already in the `VoterList` as well, so have already iterated over all of them.